### DISCUSSION TITLE
No deserialization found for type dynamic

### DISCUSSION THREAD

**Discussion by dongri727** [serverpod/serverpod#1554] [Q&A] [answered]
Please tell me.
I am trying to get one element from an intermediate table for many-to-many, eliminating duplicates.
I have written the Endpoint as follows.
```dart
  Future<List<dynamic>> getPPlaceNarrowed(Session session) async {
    var placeInvIds = await session.db.query(
      'SELECT DISTINCT place_id FROM principal_place'
    );

    return placeInvIds.map((row) => row[2]).toList();
  }
```

From the Flutter side, we call it as follows.
```dart
  List<dynamic> placeInvIds = [];
  List<Places> listPlaceInv = [];
  final List<String> filtersPlaceInv = <String>[];
  final List<int> filtersPlaceInvId = <int>[];

  fetchPlaceInvolvedLookedFor() async {
    try {
      placeInvIds = await client.principalPlace.getPPlaceNarrowed();  //HERE
      listPlaceInv = await client.places.getPlacesInv(placeInvIds: placeInvIds);
      print(listPlaceInv);
      notifyListeners();
    } catch (e) {
      debugPrint('$e');
    }
  }
```

In response, the following Error is generated and the call cannot be made.
flutter: Failed call: principalPlace.getPPlaceNarrowed
flutter: FormatException: No deserialization found for type dynamic

What can I do with this ? Thank you.

**Comment by SandPod** [comment]
Hey! 👋 

Unfortunately Serverpod doesn't support deserializing a `dynamic` type. But this should be easily fixed.

Do you now what type `place_id` is?
If so, instead of having the endpoint return a `Future<List<dynamic>>` you could specify the type that you would like to return.

Say for example that `place_id` is an int, then your return type would be `Future<List<int>>`.

**Comment by dongri727** [comment]
Thank you for your instruction.

I have tried the following modification, but it does not work with the following Error.
Sorry to be so basic, but I would appreciate it if you could tell me how to fix it.
I asked ChatGPT but they could not fix it.

```dart
  Future<List<int>>> getPPlaceNarrowed(Session session) async {
    var placeInvIds = await session.db
        .query('SELECT DISTINCT place_id FROM principal_place');

    return placeInvIds.map((row) => row.place_id).toList(); 
  }
```

The getter 'place_id' isn't defined for the type 'List<dynamic>'.
Try importing the library that defines 'place_id', correcting the name to the name of an existing getter, or defining a getter or field named 'place_id'.

The calibration of the relevant table is as follows
```yaml
class: PrincipalPlace
table: principal_place
fields: 
  principal_id: int
  place_id: int
```

And if I want to do many-to-many searches, should I wait for Serverpod 1.2?
When will the upgrade be available?

Translated with www.DeepL.com/Translator (free version)

**Reply by SandPod** [reply, marked as answer]
Since you are running this as a raw query `await session.db.query` the result won't be a desrialized type that you can call methods on. But instead will be `A [List] of rows represented of another [List] with columns...`.
  
  If I understand your initial query correctly, this should give you the place ids as a list to the client:
  ``` dart
    Future<List<int>> getPPlaceNarrowed(Session session) async {
    var placeInvIds = await session.db.query(
      'SELECT DISTINCT place_id FROM principal_place'
    );

    return placeInvIds.map((row) => row[0]).toList();
  }
  ```
  
  I use `row[0]` because it only looks like you are only selecting a single column in your query.
  
  Another option is to use our typed query system to do the same request to return the full type stored in the database, note that this would fetch all rows and not remove duplicate `place_id` rows.

  ``` dart
    Future<List<PrincipalPlace>> getPPlaceNarrowed(Session session) async {
    return await PrincipalPlace.db.find(
        session,
    );
  }
  ```

or if you only want to return the `place_id`:


  ``` dart
    Future<List<int>> getPPlaceNarrowed(Session session) async {
    var place = await PrincipalPlace.db.find(
        session,
    );

    return place.map((p) => p.place_id).toList();
  }
  ```
  
  
  ### Many to many
  Yes, many-to-many relation queries will be supported in Serverpod 1.2.
